src/Microsoft.Health.Fhir.SqlServer/Features/Search/CustomQueries.cs

```diff
@@ -0,0 +1,55 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Data.SqlClient;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Core;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search
+{
+    internal static class CustomQueries
+    {
+        private static DateTimeOffset _lastUpdatedQueryCache = DateTimeOffset.MinValue;
+        public static readonly Dictionary<string, string> QueryStore = new Dictionary<string, string>();
+
+        public static string CheckQueryHash(SqlConnection connection, string hash, ILogger<SqlServerSearchService> logger)
+        {
+            var now = Clock.UtcNow;
+            if (now > _lastUpdatedQueryCache.AddMinutes(1))
+            {
+                using (SqlCommand sqlCommand = connection.CreateCommand()) // WARNING, this code will not set sqlCommand.Transaction. Sql transactions via C#/.NET are not supported in this method.
+                {
+                    try
+                    {
+                        sqlCommand.CommandText = "SELECT ROUTINE_NAME FROM INFORMATION_SCHEMA.ROUTINES WHERE ROUTINE_TYPE = 'PROCEDURE' AND ROUTINE_NAME like 'CustomQuery_%'";
+                        using (SqlDataReader reader = sqlCommand.ExecuteReader())
+                        {
+                            if (reader.Read())
+                            {
+                                string sprocName = reader.GetString(0);
+                                var tokens = sprocName.Split('_');
+                                if (tokens.Length == 2)
+                                {
+                                    QueryStore.Add(tokens[1], sprocName);
```

Comment:
This doesn't look thread-safe in this context

Reply:
If we use Tom's retriable reader, we would be able to get entire list, and then swap it with existing one under lock.

Reply:
Changed to concurrent dictionary, and also swap entire dictionary after successful query.

Reply:
If you build entire dictionary and swap, there is no need to have it concurrent.